stats/exporter/stackdriver/stackdriver.go

```diff
@@ -0,0 +1,242 @@
+	// the backend.
+	// Optional.
+	ExportDelayThreshold time.Duration
+
+	// ExportCountThreshold determines how many view data events
+	// can be buffered before batch uploading them to the backend.
+	// Optional.
+	ExportCountThreshold int
+
+	bundler *bundler.Bundler
+
+	measuresMu sync.Mutex
+	measures   map[stats.Measure]struct{} // measures already created remotely
+
+	once sync.Once
+	c    *monitoring.MetricClient
+}
+
+// Export exports to the Stackdriver Monitoroing if view data
+// has one or more rows.
+func (e *Exporter) Export(vd *stats.ViewData) {
+	e.once.Do(e.newClient)
+	if len(vd.Rows) == 0 {
+		return
+	}
+	e.bundler.Add(vd, 1)
+}
+
+func (e *Exporter) onError(err error) {
+	if e.OnError != nil {
+		e.OnError(err)
+		return
+	}
+	log.Printf("Failed to export to Stackdriver Monitoring: %v", err)
+}
+
+func (e *Exporter) newClient() {
+	client, err := monitoring.NewMetricClient(context.Background(), e.ClientOptions...)
+	if err != nil {
+		e.OnError(err)
+		return
+	}
+	e.c = client
+	e.measures = make(map[stats.Measure]struct{})
+	e.bundler = bundler.NewBundler((*stats.ViewData)(nil), func(bundle interface{}) {
+		vds := bundle.([]*stats.ViewData)
+		if err := e.upload(vds); err != nil {
+			e.onError(err)
+		}
+	})
+	e.bundler.DelayThreshold = e.ExportDelayThreshold
+	e.bundler.BundleCountThreshold = e.ExportCountThreshold
+}
+
+func (e *Exporter) upload(vds []*stats.ViewData) error {
+	ctx := context.Background()
+
+	// TODO(jbd): Create measures concurrently?
+	for _, vd := range vds {
+		if err := e.createMeasure(ctx, vd); err != nil {
+			return err
+		}
+	}
+	if err := e.c.CreateTimeSeries(ctx, e.makeReq(vds)); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (e *Exporter) makeReq(vds []*stats.ViewData) *monitoringpb.CreateTimeSeriesRequest {
+	timeSeries := make([]*monitoringpb.TimeSeries, len(vds))
+	for i, vd := range vds {
+		timeSeries[i] = &monitoringpb.TimeSeries{
+			Metric: &metricpb.Metric{
+				Type: path.Join("custom.googleapis.com", vd.V.Measure().Name()),
+				// TODO(jbd): Add labels.
+			},
+			Resource: &monitoredrespb.MonitoredResource{
+				Type:   "global",
+				Labels: map[string]string{"project_id": e.ProjectID},
+			},
+			Points: newPoints(vd),
+		}
+	}
+	return &monitoringpb.CreateTimeSeriesRequest{
+		Name:       monitoring.MetricProjectPath(e.ProjectID),
+		TimeSeries: timeSeries,
+	}
+}
+
+func (e *Exporter) createMeasure(ctx context.Context, vd *stats.ViewData) error {
+	e.measuresMu.Lock()
+	defer e.measuresMu.Unlock()
+
+	m := vd.V.Measure()
+	agg := vd.V.Aggregation()
+
+	_, ok := e.measures[m]
+	if ok {
+		return nil
+	}
+
+	var metricKind metricpb.MetricDescriptor_MetricKind
+	var valueType metricpb.MetricDescriptor_ValueType
+
+	switch agg.(type) {
+	case stats.CountAggregation:
+		metricKind = metricpb.MetricDescriptor_CUMULATIVE
+		valueType = metricpb.MetricDescriptor_INT64
+	case stats.DistributionAggregation:
+		metricKind = metricpb.MetricDescriptor_GAUGE
```

Comment:
Should this match `View.Window`? i.e CumulativeWindow -> CUMULATIVE, SlidingCountWindow -> DELTA

Reply:
I just tried with the API Explorer (https://developers.google.com/apis-explorer/#p/monitoring.googleapis.com/v3/monitoring.projects.metricDescriptors.create), it seems DELTA is not supported for custom metric :( For Java we don't support exporting interval/sliding window views in this version.